---
title: "Code for generating an instance of iSEE for the thyroid maturation dataset (Romitti et al. 2021)"
author:
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
date: "`r BiocStyle::doc_date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Retrieving the data

Retrieving the datasets from the GEO repository

```
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4987nnn/GSM4987904/suppl/GSM4987904%5Fbarcodes%2Etsv%2Egz -O GSM4987904/barcodes.tsv.gz
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4987nnn/GSM4987904/suppl/GSM4987904%5Ffeatures%2Etsv%2Egz -O GSM4987904/features.tsv.gz
wget https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4987nnn/GSM4987904/suppl/GSM4987904%5Fmatrix%2Emtx%2Egz -O GSM4987904/matrix.mtx.gz
```


```{r}
library(SingleCellExperiment)
library(iSEE)
library(scater)
library(scran)
library(scuttle)
library(DropletUtils)
library(BiocSingular)
library(batchelor)
library(SingleR)
library(celldex)
library(scDblFinder)
library(Matrix)
library(bluster)
library(dplyr)
```


# Going DIY first

```{r}
sce <- read10xCounts("GSM4987904")

sce

sce$sample_id <- "GSM4987904"

pryr::object_size(sce)

rowData(sce)
head(rownames(sce))
rownames(sce) <- uniquifyFeatureNames(rowData(sce)$ID, rowData(sce)$Symbol)
head(rownames(sce))

library(AnnotationHub)
ah <- AnnotationHub()
query(ah, c("musculus", "Ensembl", "EnsDb"))
ens.mm.v102 <- ah[["AH89211"]]
genes(ens.mm.v102)[,2]

is.mito <- grepl("^mt-", rownames(sce))

chr.loc <- mapIds(ens.mm.v102, keys=rownames(sce),
                  keytype="GENENAME", column="SEQNAME")
is.mito <- which(chr.loc=="MT")

ref_annot_immgen <- ImmGenData()
# ref_annot_novershtern <- NovershternHematopoieticData()
# ref_annot_dice <- DatabaseImmuneCellExpressionData()
# ref_annot_monaco <- MonacoImmuneData()
# ref_annot_mouse <- MouseRNAseqData()
# ref_annot_encode <- BlueprintEncodeData()

rowData(sce)$location <- chr.loc

sce <- addPerFeatureQC(sce)

rowData(sce)

sce <- addPerCellQC(sce, subsets=list(Mito=is.mito))
qcstats <- perCellQCMetrics(sce, subsets=list(Mito=is.mito))
filtered <- quickPerCellQC(qcstats, percent_subsets="subsets_Mito_percent")
filtered
colSums(as.data.frame(filtered))
table(filtered$low_n_features, filtered$high_subsets_Mito_percent)

sce$discard <- filtered$discard

plotColData(sce, y="subsets_Mito_percent", colour_by="discard")
plotColData(sce, y="sum", colour_by="discard")
plotColData(sce, y="detected", colour_by="discard")

plotColData(sce, x="sum", y="subsets_Mito_percent", colour_by="discard")
plotColData(sce, x="detected", y="subsets_Mito_percent", colour_by="discard")

sce <- sce[, !(filtered$discard | sce$subsets_Mito_percent > 50)]
sce

clusters <- scran::quickCluster(sce, min.size = 50)
sce <- scran::computeSumFactors(sce, clusters=clusters)
sce <- scater::logNormCounts(sce)

dec <- scran::modelGeneVar(sce)
hvgs <- scran::getTopHVGs(dec, prop = 0.1)

set.seed(42)
sce <- scater::runPCA(sce, subset_row=hvgs)

clusters <- clusterRows(reducedDim(sce, "PCA"), NNGraphParam())

sce <- scater::runTSNE(sce, dimred="PCA")
sce <- scater::runUMAP(sce, dimred = 'PCA', external_neighbors=TRUE)

snn.gr <- buildSNNGraph(sce, use.dimred="PCA", k=25)
sce$cluster_round1 <- factor(igraph::cluster_louvain(snn.gr)$membership)

plotTSNE(sce, colour_by="cluster_round1", text_by="cluster_round1")

# Doublet detection
sce <- scDblFinder(sce)
table(sce$scDblFinder.class)

plotTSNE(sce, colour_by="scDblFinder.score", text_by="cluster_round1")
plotTSNE(sce, colour_by="scDblFinder.class", text_by="cluster_round1")

# cell type annotation, round one!
pred_celltypes <- SingleR(test = sce, 
                          ref = ref_annot_immgen, 
                          labels = ref_annot_immgen$label.main,
                          BPPARAM = BiocParallel::MulticoreParam(6))
table(pred_celltypes$labels)
sce$pred_celltypes <- pred_celltypes$labels
plotTSNE(sce, colour_by= "pred_celltypes", text_by = "pred_celltypes")
plotUMAP(sce, colour_by= "pred_celltypes", text_by = "pred_celltypes")

# pred_celltypes_novershtern <- SingleR(test = sce, 
#                                       ref = ref_annot_novershtern, 
#                                       labels = ref_annot_novershtern$label.main,
#                                       BPPARAM = BiocParallel::MulticoreParam(6))
# table(pred_celltypes_novershtern$labels)
# sce$pred_celltypes_novershtern <- pred_celltypes_novershtern$labels
# plotTSNE(sce, colour_by= "pred_celltypes_novershtern", text_by = "pred_celltypes_novershtern")

cowplot::plot_grid(
  plotlist = list(
    plotUMAP(sce, colour_by= "Pax8"),
    plotUMAP(sce, colour_by= "Tg"),
    plotUMAP(sce, colour_by= "Acta2"),
    plotUMAP(sce, colour_by= "Pou5f1"),
    plotUMAP(sce, colour_by= "Sox2"),
    plotUMAP(sce, colour_by= "Pax6"),
    plotUMAP(sce, colour_by= "Krt5"),
    plotUMAP(sce, colour_by= "Col1a2"),
    plotUMAP(sce, colour_by= "Perp"),
    plotUMAP(sce, colour_by= "Fcer1g"),
    plotUMAP(sce, colour_by= "Flt1"),
    plotUMAP(sce, colour_by= "Pou3f3"),
    plotUMAP(sce, colour_by= "Hand1")
  )
)

sce_diy <- sce
```

# Following the workflow from the authors

... as available in https://github.com/selifeski/mouse-thyroid-invitro

```{r}
# Load the data
library("Seurat")
rawdata <- Read10X(data.dir = "GSM4987904/")
MmThyroid <- CreateSeuratObject(counts = rawdata,
                                project = "thyroid_data",
                                min.cells = 3, min.features = 200, 
                                names.field = 2, names.delim = "-")
MmThyroid
```

**Quality Control**

Some information about mitochondrial genes, which are important for quality control.

```{r}
MmThyroid[["percent.mt"]] <- PercentageFeatureSet(MmThyroid, pattern = "^mt-")
```

A violin plot of some of the computed quality measures:

  - the number of genes expressed in the count matrix
  - the total counts per cell
  - the percentage of counts in mitochondrial genes

```{r}
VlnPlot(MmThyroid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,
        pt.size = 0, group.by = "orig.ident")
```

Remove cells that have high mitochondrial gene percentage or too many/too low total counts.

```{r}
MmThyroid <- subset(MmThyroid, subset = nFeature_RNA > 800 & nFeature_RNA < 5000
                    & percent.mt < 20 & nCount_RNA > 2000 & nCount_RNA < 20000)
```

**After filtration**

```{r}
VlnPlot(MmThyroid, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3,
        pt.size = 0, group.by = "orig.ident")
```

**Function to convert human genes to mouse**

```{r echo=T, results='hide', message=FALSE, warning=FALSE}
library(biomaRt)
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
human2ms <- function(human.genes){
  genesV2 = getLDS(attributes = c("hgnc_symbol"), 
                   filters = "hgnc_symbol", 
                   values = human.genes, 
                   mart = human, 
                   attributesL = c("mgi_symbol"), 
                   martL = mouse, uniqueRows=F)
}
```

**Convert human cell cycle genes from Seurat to mouse orthologs**

```{r echo=T, message=FALSE, warning=FALSE}
## Convert cc genes of seurat to mouse
# cc.genes.updated.2019
cc.genes.updated.2019_ms <- NULL
hs2ms_s.genes = human2ms(cc.genes.updated.2019$s.genes)
cc.genes.updated.2019_ms$s.genes <- unique(hs2ms_s.genes[, 2])
hs2ms_g2m.genes = human2ms(cc.genes.updated.2019$g2m.genes)
cc.genes.updated.2019_ms$g2m.genes <- unique(hs2ms_g2m.genes[, 2])
cc.genes.updated.2019_ms
```

**Cell cycle scoring**

```{r echo=T, results='hide',message=FALSE, warning=FALSE}
MmThyroid <- CellCycleScoring(MmThyroid, s.features = cc.genes.updated.2019_ms$s.genes,
                              g2m.features = cc.genes.updated.2019_ms$g2m.genes)
```

**Normalizing the data**

Total count normalize and logaritmize the data. And identify highly variable genes. 2,000 features were selected for downstream PCA.

```{r echo=T, results='hide', message=FALSE, warning=FALSE}
MmThyroid <- SCTransform(MmThyroid,
                         vars.to.regress = c("nCount_RNA","percent.mt","S.Score", "G2M.Score"),
                         variable.features.n = 2000)
```


## Dimensional Reduction

Reduce the dimensionality of the data. 

```{r echo=T, results='hide'}
MmThyroid <- RunPCA(MmThyroid, features = VariableFeatures(object = MmThyroid))
```

The contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells. Here I chose to use the first 10 PCs based on the variety of clusters.

```{r}
# Determine the dimensionality of the dataset
ElbowPlot(MmThyroid, ndims=30)
dims.use = 1:10
```

```{r echo=T, results='hide', message=FALSE, warning=FALSE}
# Run non-linear dimensional reduction
MmThyroid <- RunUMAP(MmThyroid, dims = dims.use)
# Cluster the cells
MmThyroid <- FindNeighbors(MmThyroid, dims = dims.use)
MmThyroid <- FindClusters(MmThyroid, resolution = 0.7)
```

**Visualize the clusters**

```{r}
DimPlot(MmThyroid, reduction = "umap", pt.size = 1, label = F, label.size = 6)
```

## Transfer cell types, umap coordinates and leiden clusters from scanpy

```{r}
metadata <- read.csv('https://raw.githubusercontent.com/selifeski/mouse-thyroid-invitro/master/files/metadata.csv', row.names = 1)
```

**Transfer cell types**

```{r}
cells <- rownames(metadata)
cell_types <- metadata$cells
MmThyroid <- subset(MmThyroid, cells = cells)
MmThyroid <- AddMetaData(object = MmThyroid, metadata = cell_types, col.name = 'cells')
Idents(MmThyroid) <- MmThyroid$cells
```

**Transfer leiden clusters**

```{r}
leiden_clusters = metadata$leiden
MmThyroid <- AddMetaData(object = MmThyroid, metadata = leiden_clusters, col.name = 'leiden')
```

**Transfer UMAP coordinates**

```{r}
umap_1 = metadata$X_umap.0
umap_2 = metadata$X_umap.1
MmThyroid <- AddMetaData(object = MmThyroid, metadata = umap_1, col.name = 'UMAP_1')
MmThyroid <- AddMetaData(object = MmThyroid, metadata = umap_2, col.name = 'UMAP_2')
MmThyroid@reductions$umap@cell.embeddings[,1] <- MmThyroid@meta.data$UMAP_1
MmThyroid@reductions$umap@cell.embeddings[,2] <- MmThyroid@meta.data$UMAP_2
```

**Visualize the final plot**
```{r}
DimPlot(MmThyroid, reduction = "umap", pt.size = 1, label = F, label.size = 6, group.by = c("cells"))
```

## Selecting mature and immature thyrocytes

**Mature cells expressing Slc5a5 and Tpo:**
```{r}
mature.cells <- which((as.numeric(MmThyroid@assays$SCT@data["Slc5a5",]) > 0.5 |
                         as.numeric(MmThyroid@assays$SCT@data["Tpo",]) > 0.5) &
                        (as.factor(MmThyroid@meta.data$cells) == 'Thyrocytes' |
                           as.factor(MmThyroid@meta.data$cells) == 'Dox-responsive cluster'))
```

**Immature cells expressing Tg, Slc5a5 and Tpo:**

```{r}
inter.cells <- which((as.numeric(MmThyroid@assays$SCT@data["Tg",]) >0.5 |
                       as.numeric(MmThyroid@assays$SCT@data["Slc5a5",]) > 0.5 |
                       as.numeric(MmThyroid@assays$SCT@data["Tpo",]) > 0.5) &
                       (as.factor(MmThyroid@meta.data$cells) == 'Thyrocytes' |
                          as.factor(MmThyroid@meta.data$cells) == 'Dox-responsive cluster'))
```

**Cells express early thyrocyte lineage markers: Nkx2-1, Pax8, Foxe1 and Hhex:**

```{r}
early.cells <- which((as.numeric(MmThyroid@assays$SCT@data["Nkx2-1",]) > 0.5 |
                       as.numeric(MmThyroid@assays$SCT@data["Pax8",]) > 0.5 |
                       as.numeric(MmThyroid@assays$SCT@data["Foxe1",]) > 0.5 |
                       as.numeric(MmThyroid@assays$SCT@data["Hhex",]) > 0.5 ) &
                       (as.factor(MmThyroid@meta.data$cells) == 'Thyrocytes' |
                          as.factor(MmThyroid@meta.data$cells) == 'Dox-responsive cluster'))
```

**Grouping mesodermal and fibroblast cells:**

```{r}
mes.cells <- which(as.factor(MmThyroid@meta.data$cells) == "Mesoderm" |
                     as.factor(MmThyroid@meta.data$cells) == "Fibroblasts")
```

**Adding cluster information to metadata:**

```{r}
MmThyroid@meta.data$cluster <- "Others"
MmThyroid@meta.data$cluster[mes.cells] <- "Mesoderm/Fibroblast"
MmThyroid@meta.data$cluster[early.cells] <- "Dox-responsive cluster"
MmThyroid@meta.data$cluster[inter.cells] <- "Immature Thyrocytes"
MmThyroid@meta.data$cluster[mature.cells] <- "Mature Thyrocytes"
```

**Visualize the cluster:**

```{r}
DimPlot(MmThyroid, group.by = "cluster", label = TRUE, repel = TRUE, pt.size = 1)+ NoLegend()
```

Save the cell information for mature and immature thyrocyte:

```{r}
# write.csv(MmThyroid@meta.data['cluster'], 'Mature_immature_early_thyrocytes.csv')
```

**Switch to Scanpy here**

...


# Entering the SingleCellExperiment universe!

Trying first with some existing functions...

```{r}
MmThyroid.sce <- as.SingleCellExperiment(MmThyroid,assay = "SCT")

sceasy::convertFormat(MmThyroid, from="seurat", to="anndata",
                       outFile='MmThyroid.h5ad')

iSEE(MmThyroid.sce)
```

Proceeding somewhat in a more manual way, where more control is ensured.

```{r}
x <- MmThyroid

assay <- Assays(object = x)

experiments <- list()
for (assayn in assay) {
  assays = list(counts = GetAssayData(object = x, 
                                      assay = assayn, 
                                      slot = "counts"), 
                logcounts = GetAssayData(object = x, 
                                         assay = assayn, 
                                         slot = "data"))
  assays <- assays[sapply(X = assays, FUN = nrow) != 0]
  sume <- SummarizedExperiment::SummarizedExperiment(assays = assays)
  experiments[[assayn]] <- sume
}
sce <- as(object = experiments[[1]], Class = "SingleCellExperiment")
sce <- SingleCellExperiment::SingleCellExperiment(sce, altExps = experiments)
orig.exp.name <- names(x = experiments[1])

sce <- SingleCellExperiment::swapAltExp(x = sce, name = orig.exp.name, 
                                        saved = NULL)

metadata <- x[[]]
metadata$ident <- Idents(object = x)
SummarizedExperiment::colData(x = sce) <- S4Vectors::DataFrame(metadata)

sce <- sce[rownames(sce) %in% rownames(experiments$SCT)]

assays(sce)[["sctcounts"]] <- assays(experiments$SCT)[["logcounts"]]

# for (assayn in assay) {
#     sce <- SingleCellExperiment::swapAltExp(x = sce, name = assayn,
#         saved = orig.exp.name)
#     SummarizedExperiment::rowData(x = sce) <- S4Vectors::DataFrame(x[[assayn]][[]])
#     sce <- SingleCellExperiment::swapAltExp(x = sce, name = orig.exp.name,
#         saved = assayn)
# }

for (dr in Seurat:::FilterObjects(object = x, classes.keep = "DimReduc")) {
  assay.used <- DefaultAssay(object = x[[dr]])
  # if (assay.used %in% SingleCellExperiment::altExpNames(x = sce)) {
    # sce <- SingleCellExperiment::swapAltExp(x = sce, 
                                            # name = assay.used, saved = orig.exp.name)
  SingleCellExperiment::reducedDim(x = sce, toupper(x = dr)) <- Embeddings(object = x[[dr]])
  # sce <- SingleCellExperiment::swapAltExp(x = sce, 
                                            # name = orig.exp.name, saved = assay.used)
  # }
}

rowData(sce)$gene_name <- rownames(sce)
rowData(sce)$chr_loc <- mapIds(ens.mm.v102, keys=rownames(sce),
                  keytype="GENENAME", column="SEQNAME")

rowData(sce)$gene_id <- mapIds(ens.mm.v102, keys=rownames(sce),
                  keytype="GENENAME", column="GENEID")

    
    # reducedDim(sce, "PCA") <- Embeddings(object = x[["pca"]]) 
    # reducedDim(sce, "UMAP") <- Embeddings(object = x[["umap"]])
    
    
    # sce <- SingleCellExperiment::swapAltExp(x = sce, name = orig.exp.name, 
        # saved = NULL)
    # return(sce)

saveRDS(sce, file = "sce_seurat_thyroid.rds")
```

# Finally iSEE!

```{r}
sce_seurat_thyroid <- readRDS("~/Development/iSEE_instances/iSEE_thyroid_organoid/sce_seurat_thyroid.rds")
```


```{r}
initial <- list()

################################################################################
# Settings for Reduced dimension plot 1
################################################################################

initial[["ReducedDimensionPlot1"]] <- new("ReducedDimensionPlot", Type = "UMAP", XAxis = 1L, YAxis = 2L, 
    FacetRowByColData = "orig.ident", FacetColumnByColData = "orig.ident", 
    ColorByColumnData = "orig.ident", ColorByFeatureNameAssay = "sctcounts", 
    ColorBySampleNameColor = "#FF0000", ShapeByColumnData = "orig.ident", 
    SizeByColumnData = "nCount_RNA", FacetRowBy = "None", FacetColumnBy = "None", 
    ColorBy = "Feature name", ColorByDefaultColor = "#000000", 
    ColorByFeatureName = "Tg", ColorByFeatureSource = "RowDataTable1", 
    ColorByFeatureDynamicSource = FALSE, ColorBySampleName = "AAACCCAAGCCTATCA-1", 
    ColorBySampleSource = "---", ColorBySampleDynamicSource = FALSE, 
    ShapeBy = "None", SizeBy = "None", SelectionAlpha = 0.1, 
    ZoomData = numeric(0), BrushData = list(), VisualBoxOpen = TRUE, 
    VisualChoices = "Color", ContourAdd = FALSE, ContourColor = "#0000FF", 
    PointSize = 1, PointAlpha = 1, Downsample = FALSE, DownsampleResolution = 200, 
    CustomLabels = FALSE, CustomLabelsText = "AAACCCAAGCCTATCA-1", 
    FontSize = 1, LegendPointSize = 1, LegendPosition = "Bottom", 
    HoverInfo = TRUE, LabelCenters = FALSE, LabelCentersBy = "orig.ident", 
    LabelCentersColor = "#000000", VersionInfo = list(iSEE = structure(list(
        c(2L, 5L, 0L)), class = c("package_version", "numeric_version"
    ))), PanelId = c(ReducedDimensionPlot = 1L), PanelHeight = 500L, 
    PanelWidth = 3L, SelectionBoxOpen = FALSE, RowSelectionSource = "---", 
    ColumnSelectionSource = "---", DataBoxOpen = FALSE, RowSelectionDynamicSource = FALSE, 
    ColumnSelectionDynamicSource = FALSE, RowSelectionRestrict = FALSE, 
    ColumnSelectionRestrict = FALSE, SelectionHistory = list())

################################################################################
# Settings for Row data table 1
################################################################################

initial[["RowDataTable1"]] <- new("RowDataTable", Selected = "Tg", Search = "Tg", SearchColumns = c("", 
"", ""), HiddenColumns = character(0), VersionInfo = list(iSEE = structure(list(
    c(2L, 5L, 0L)), class = c("package_version", "numeric_version"
))), PanelId = c(RowDataTable = 1L), PanelHeight = 500L, PanelWidth = 3L, 
    SelectionBoxOpen = FALSE, RowSelectionSource = "---", ColumnSelectionSource = "---", 
    DataBoxOpen = FALSE, RowSelectionDynamicSource = FALSE, ColumnSelectionDynamicSource = FALSE, 
    RowSelectionRestrict = FALSE, ColumnSelectionRestrict = FALSE, 
    SelectionHistory = list())

################################################################################
# Settings for Feature assay plot 1
################################################################################

initial[["FeatureAssayPlot1"]] <- new("FeatureAssayPlot", Assay = "sctcounts", XAxis = "Column data", 
    XAxisColumnData = "cluster", XAxisFeatureName = "Xkr4", XAxisFeatureSource = "---", 
    XAxisFeatureDynamicSource = FALSE, YAxisFeatureName = "Tg", 
    YAxisFeatureSource = "RowDataTable1", YAxisFeatureDynamicSource = FALSE, 
    FacetRowByColData = "orig.ident", FacetColumnByColData = "orig.ident", 
    ColorByColumnData = "orig.ident", ColorByFeatureNameAssay = "logcounts", 
    ColorBySampleNameColor = "#FF0000", ShapeByColumnData = "orig.ident", 
    SizeByColumnData = "nCount_RNA", FacetRowBy = "None", FacetColumnBy = "None", 
    ColorBy = "None", ColorByDefaultColor = "#000000", ColorByFeatureName = "Xkr4", 
    ColorByFeatureSource = "---", ColorByFeatureDynamicSource = FALSE, 
    ColorBySampleName = "AAACCCAAGCCTATCA-1", ColorBySampleSource = "---", 
    ColorBySampleDynamicSource = FALSE, ShapeBy = "None", SizeBy = "None", 
    SelectionAlpha = 0.1, ZoomData = numeric(0), BrushData = list(), 
    VisualBoxOpen = FALSE, VisualChoices = "Color", ContourAdd = FALSE, 
    ContourColor = "#0000FF", PointSize = 1, PointAlpha = 1, 
    Downsample = FALSE, DownsampleResolution = 200, CustomLabels = FALSE, 
    CustomLabelsText = "AAACCCAAGCCTATCA-1", FontSize = 1, LegendPointSize = 1, 
    LegendPosition = "Bottom", HoverInfo = TRUE, LabelCenters = FALSE, 
    LabelCentersBy = "orig.ident", LabelCentersColor = "#000000", 
    VersionInfo = list(iSEE = structure(list(c(2L, 5L, 0L)), class = c("package_version", 
    "numeric_version"))), PanelId = c(FeatureAssayPlot = 1L), 
    PanelHeight = 500L, PanelWidth = 3L, SelectionBoxOpen = FALSE, 
    RowSelectionSource = "---", ColumnSelectionSource = "---", 
    DataBoxOpen = FALSE, RowSelectionDynamicSource = FALSE, ColumnSelectionDynamicSource = FALSE, 
    RowSelectionRestrict = FALSE, ColumnSelectionRestrict = FALSE, 
    SelectionHistory = list())

################################################################################
# Settings for Complex heatmap 1
################################################################################

initial[["ComplexHeatmapPlot1"]] <- new("ComplexHeatmapPlot", Assay = "sctcounts", CustomRows = TRUE, 
    CustomRowsText = "Pax8\nTg\nActa2\nPou5f1\nSox2\nKrt5\nCol1a2\nPerp\nFcer1g\nFlt1\nPou3f3\n", 
    ClusterRows = FALSE, ClusterRowsDistance = "spearman", ClusterRowsMethod = "ward.D2", 
    DataBoxOpen = FALSE, VisualChoices = "Annotations", ColumnData = "cluster", 
    RowData = character(0), CustomBounds = FALSE, LowerBound = NA_real_, 
    UpperBound = NA_real_, AssayCenterRows = FALSE, AssayScaleRows = FALSE, 
    DivergentColormap = "purple < black < yellow", ShowDimNames = "Rows", 
    LegendPosition = "Bottom", LegendDirection = "Horizontal", 
    VisualBoxOpen = FALSE, NamesRowFontSize = 10, NamesColumnFontSize = 10, 
    ShowColumnSelection = TRUE, OrderColumnSelection = TRUE, 
    VersionInfo = list(iSEE = structure(list(c(2L, 5L, 0L)), class = c("package_version", 
    "numeric_version"))), PanelId = c(ComplexHeatmapPlot = 1L), 
    PanelHeight = 500L, PanelWidth = 3L, SelectionBoxOpen = FALSE, 
    RowSelectionSource = "---", ColumnSelectionSource = "---", 
    RowSelectionDynamicSource = FALSE, ColumnSelectionDynamicSource = FALSE, 
    RowSelectionRestrict = FALSE, ColumnSelectionRestrict = FALSE, 
    SelectionHistory = list())
```

```{r}
iSEE(sce_seurat_thyroid, initial = initial, appTitle = "iSEE - Thyroid Maturation")
```



